// rmw-shell/lib/containers/Root/Root.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactRedux = require('react-redux');

var _locales = require('../../config/locales');

var _locales2 = _interopRequireDefault(_locales);

var _themes = require('../../config/themes');

var _themes2 = _interopRequireDefault(_themes);

var _styles = require('@material-ui/core/styles');

var _MuiThemeProvider = require('@material-ui/core/styles/MuiThemeProvider');

var _MuiThemeProvider2 = _interopRequireDefault(_MuiThemeProvider);

var _MuiPickersUtilsProvider = require('material-ui-pickers/utils/MuiPickersUtilsProvider');

var _MuiPickersUtilsProvider2 = _interopRequireDefault(_MuiPickersUtilsProvider);

var _momentUtils = require('material-ui-pickers/utils/moment-utils');

var _momentUtils2 = _interopRequireDefault(_momentUtils);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _reactIntl = require('react-intl');

var _AppLayout = require('../../containers/AppLayout');

var _AppLayout2 = _interopRequireDefault(_AppLayout);

var _firekit = require('firekit');

var _createBrowserHistory = require('history/createBrowserHistory');

var _createBrowserHistory2 = _interopRequireDefault(_createBrowserHistory);

var _reactRouterDom = require('react-router-dom');

var _messaging = require('../../utils/messaging');

var _actions = require('../../store/persistentValues/actions');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var history = (0, _createBrowserHistory2.default)();

var Root = function (_Component) {
  _inherits(Root, _Component);

  function Root() {
    var _temp, _this, _ret;

    _classCallCheck(this, Root);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _Component.call.apply(_Component, [this].concat(args))), _this), _this.handlePresence = function (user, firebaseApp) {

      var myConnectionsRef = firebaseApp.database().ref('users/' + user.uid + '/connections');

      var lastOnlineRef = firebaseApp.database().ref('users/' + user.uid + '/lastOnline');
      lastOnlineRef.onDisconnect().set(new Date());

      var con = myConnectionsRef.push(true);
      con.onDisconnect().remove();
    }, _this.onAuthStateChanged = function (user, firebaseApp) {
      var _this$props = _this.props,
          clearInitialization = _this$props.clearInitialization,
          watchConnection = _this$props.watchConnection,
          watchList = _this$props.watchList,
          watchPath = _this$props.watchPath,
          appConfig = _this$props.appConfig;


      clearInitialization();

      if (user) {

        _this.handlePresence(user, firebaseApp);
        setTimeout(function () {
          watchConnection(firebaseApp);
        }, 1000);

        var userData = {
          displayName: user.displayName ? user.displayName : 'UserName',
          email: user.email ? user.email : ' ',
          photoURL: user.photoURL,
          emailVerified: user.emailVerified,
          isAnonymous: user.isAnonymous,
          uid: user.uid,
          providerData: user.providerData
        };

        var publicProviderData = [];

        user.providerData.forEach(function (provider) {
          publicProviderData.push({
            providerId: provider.providerId,
            displayName: provider.displayName ? provider.displayName : null
          });
        });

        var publicUserData = {
          displayName: user.displayName ? user.displayName : 'UserName',
          photoURL: user.photoURL,
          uid: user.uid,
          providerData: publicProviderData
        };

        watchList(firebaseApp, 'user_grants/' + user.uid);
        watchPath(firebaseApp, 'admins/' + user.uid);

        if (appConfig.onAuthStateChanged) {
          try {
            appConfig.onAuthStateChanged(user, _this.props, firebaseApp);
          } catch (err) {
            console.warn(err);
          }
        }

        firebaseApp.database().ref('users/' + user.uid).update(publicUserData);

        (0, _messaging.initializeMessaging)(_extends({}, _this.props, { firebaseApp: firebaseApp, history: history, auth: userData }), true);

        return userData;
      } else {
        return null;
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  Root.prototype.componentWillMount = function componentWillMount() {
    var _this2 = this;

    var _props = this.props,
        watchAuth = _props.watchAuth,
        appConfig = _props.appConfig;


    appConfig.firebaseLoad().then(function (_ref) {
      var firebaseApp = _ref.firebaseApp;

      watchAuth(firebaseApp, function (user) {
        return _this2.onAuthStateChanged(user, firebaseApp);
      });
    });
  };

  Root.prototype.componentWillUnmount = function componentWillUnmount() {
    //const { clearApp }= this.props;
    //clearApp(this.firebaseApp); //TODO: add it after firekit update
  };

  Root.prototype.render = function render() {
    var _props2 = this.props,
        appConfig = _props2.appConfig,
        locale = _props2.locale,
        themeSource = _props2.themeSource;


    var messages = _extends({}, (0, _locales.getLocaleMessages)(locale, _locales2.default), (0, _locales.getLocaleMessages)(locale, appConfig.locales));
    var source = (0, _themes2.default)(themeSource, appConfig.themes);
    var theme = (0, _styles.createMuiTheme)(source);

    return _react2.default.createElement(
      _MuiPickersUtilsProvider2.default,
      { utils: _momentUtils2.default },
      _react2.default.createElement(
        _MuiThemeProvider2.default,
        { theme: theme },
        _react2.default.createElement(
          _reactIntl.IntlProvider,
          { locale: locale, key: locale, messages: messages },
          _react2.default.createElement(
            _reactRouterDom.Router,
            { history: history },
            _react2.default.createElement(
              _reactRouterDom.Switch,
              null,
              _react2.default.createElement(_reactRouterDom.Route, { children: function children(props) {
                  return _react2.default.createElement(_AppLayout2.default, props);
                } })
            )
          )
        )
      )
    );
  };

  return Root;
}(_react.Component);

Root.propTypes = process.env.NODE_ENV !== "production" ? {
  locale: _propTypes2.default.string.isRequired,
  themeSource: _propTypes2.default.object.isRequired
} : {};

var mapStateToProps = function mapStateToProps(state, ownProps) {
  var locale = state.locale,
      themeSource = state.themeSource,
      persistentValues = state.persistentValues,
      simpleValues = state.simpleValues;


  var notificationPermissionRequested = persistentValues.notificationPermissionRequested;

  return {
    locale: locale,
    themeSource: themeSource,
    notificationPermissionRequested: notificationPermissionRequested,
    simpleValues: simpleValues
  };
};

exports.default = (0, _reactRedux.connect)(mapStateToProps, {
  watchAuth: _firekit.watchAuth,
  clearInitialization: _firekit.clearInitialization,
  watchConnection: _firekit.initConnection,
  watchList: _firekit.watchList,
  watchPath: _firekit.watchPath,
  initMessaging: _firekit.initMessaging,
  setPersistentValue: _actions.setPersistentValue
})(Root);
module.exports = exports['default'];
