// rmw-shell/es/containers/Activity/Activity.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import AppBar from '@material-ui/core/AppBar';
import Icon from '@material-ui/core/Icon';
import IconButton from '@material-ui/core/IconButton';
import LinearProgress from '@material-ui/core/LinearProgress';
import MenuIcon from '@material-ui/icons/Menu';
import PropTypes from 'prop-types';
import React from 'react';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import classNames from 'classnames';
import drawerActions from '../../store/drawer/actions';
import withWidth, { isWidthDown } from '@material-ui/core/withWidth';
import { Helmet } from 'react-helmet';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { injectIntl } from 'react-intl';
import { withStyles } from '@material-ui/core/styles';

var drawerWidth = 240;

var styles = function styles(theme) {
  return {
    root: {
      display: 'flex',
      flexDirection: 'column',
      height: '100vh'
    },
    appBar: {
      zIndex: theme.zIndex.drawer + 1,
      transition: theme.transitions.create(['width', 'margin'], {
        easing: theme.transitions.easing.sharp,
        duration: theme.transitions.duration.leavingScreen
      })
    },
    menuButton: {
      marginLeft: 12,
      marginRight: 12
    },
    toolbar: _extends({
      alignItems: 'center',
      justifyContent: 'flex-end'
    }, theme.mixins.toolbar),
    content: {
      flex: 1,
      backgroundColor: theme.palette.background.default
    },

    appBarShift: {
      //marginLeft: drawerWidth,
      width: 'calc(100% - ' + drawerWidth + 'px)',
      transition: theme.transitions.create(['width', 'margin'], {
        easing: theme.transitions.easing.sharp,
        duration: theme.transitions.duration.enteringScreen
      })
    },
    hide: {
      display: 'none'
    },
    grow: {
      flex: '1 1 auto'
    }
  };
};

var Activity = function (_React$Component) {
  _inherits(Activity, _React$Component);

  function Activity() {
    var _temp, _this, _ret;

    _classCallCheck(this, Activity);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _React$Component.call.apply(_React$Component, [this].concat(args))), _this), _this.handleDrawerToggle = function () {
      var _this$props = _this.props,
          setDrawerMobileOpen = _this$props.setDrawerMobileOpen,
          drawer = _this$props.drawer;

      setDrawerMobileOpen(!drawer.mobileOpen);
    }, _this.handleDrawerOpen = function () {
      var setDrawerOpen = _this.props.setDrawerOpen;

      setDrawerOpen(true);
    }, _this.handleDrawerClose = function () {
      var setDrawerOpen = _this.props.setDrawerOpen;

      setDrawerOpen(false);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  Activity.prototype.render = function render() {
    var _props = this.props,
        classes = _props.classes,
        theme = _props.theme,
        children = _props.children,
        drawer = _props.drawer,
        intl = _props.intl,
        title = _props.title,
        pageTitle = _props.pageTitle,
        width = _props.width,
        appBarContent = _props.appBarContent,
        isLoading = _props.isLoading,
        onBackClick = _props.onBackClick,
        isOffline = _props.isOffline;


    var headerTitle = '';

    if (typeof title === 'string' || title instanceof String) {
      headerTitle = title;
    }

    if (pageTitle) {
      headerTitle = pageTitle;
    }

    //const smDown = width === 'sm' || width === 'xs'
    var smDown = isWidthDown('sm', width);

    return React.createElement(
      'div',
      { className: classes.root },
      React.createElement(
        Helmet,
        null,
        React.createElement('meta', { name: 'theme-color', content: theme.palette.primary.main }),
        React.createElement('meta', { name: 'apple-mobile-web-app-status-bar-style', content: theme.palette.primary.main }),
        React.createElement('meta', { name: 'msapplication-navbutton-color', content: theme.palette.primary.main }),
        React.createElement(
          'title',
          null,
          headerTitle
        )
      ),
      React.createElement(
        AppBar,
        {
          position: width !== 'sm' && width !== 'xs' ? 'absolute' : undefined,
          className: width !== 'sm' && width !== 'xs' ? classNames(classes.appBar, drawer.open && classes.appBarShift) : classes.appBar
        },
        React.createElement(
          Toolbar,
          { disableGutters: true },
          true && React.createElement(LinearProgress, null),
          React.createElement(
            IconButton,
            {
              color: 'inherit',
              'aria-label': 'open drawer',
              onClick: !drawer.open ? this.handleDrawerOpen : this.handleDrawerToggle,
              className: classNames(!smDown && classes.menuButton, drawer.open && !smDown && classes.hide, onBackClick && classes.hide)
            },
            React.createElement(MenuIcon, null)
          ),
          React.createElement(
            IconButton,
            {
              color: 'inherit',
              'aria-label': 'open drawer',
              onClick: onBackClick,
              className: classNames(!smDown && classes.menuButton, !onBackClick && classes.hide)
            },
            React.createElement(
              Icon,
              null,
              'chevron_left'
            )
          ),
          !onBackClick && drawer.open && React.createElement('div', { style: { marginRight: 32 } }),
          React.createElement(
            Typography,
            { variant: 'title', color: 'inherit', noWrap: true },
            headerTitle
          ),
          React.createElement('div', { className: classes.grow }),
          appBarContent
        )
      ),
      React.createElement('div', { className: classes.toolbar }),
      isLoading && React.createElement(LinearProgress, null),
      isOffline && React.createElement(
        'div',
        {
          style: {
            display: 'flex',
            justifyContent: 'center',
            width: '100%',
            height: 15,
            backgroundColor: theme.palette.secondary.main
          }
        },
        React.createElement(
          Typography,
          { variant: 'caption', color: 'textSecondary', noWrap: true },
          intl.formatMessage({ id: 'offline' })
        )
      ),
      React.createElement(
        'main',
        { className: classes.content },
        children
      )
    );
  };

  return Activity;
}(React.Component);

Activity.propTypes = process.env.NODE_ENV !== "production" ? {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
  drawer: PropTypes.object.isRequired
} : {};

var mapStateToProps = function mapStateToProps(state) {
  var drawer = state.drawer,
      connection = state.connection;


  return {
    drawer: drawer,
    isOffline: connection ? !connection.isConnected : false
  };
};

export default compose(connect(mapStateToProps, _extends({}, drawerActions)), withWidth(), withStyles(styles, { withTheme: true }), injectIntl)(Activity);
