// material-ui-selectable-menu-list/lib/SelectableMenuList.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _materialUi = require('material-ui');

var _List = require('material-ui/List');

var _Subheader = require('material-ui/Subheader');

var _Subheader2 = _interopRequireDefault(_Subheader);

var _Divider = require('material-ui/Divider');

var _Divider2 = _interopRequireDefault(_Divider);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function getNestedItems(hostItem, hostIndex) {
  if (hostItem.nestedItems !== undefined) {
    var items = hostItem.nestedItems.filter(function (item) {
      return item.visible !== false;
    });

    if (items.length > 0) {
      return items.map(function (item, i) {
        return getItem(item, hostIndex.toString() + i.toString());
      });
    }
  }

  return undefined;
};

function getItem(item, i) {
  delete item.visible;

  if (item !== undefined) {
    if (item.subheader !== undefined) {
      return _react2.default.createElement(
        _Subheader2.default,
        {
          key: i,
          inset: item.inset,
          style: item.style },
        item.subheader
      );
    } else if (item.divider !== undefined) {
      return _react2.default.createElement(_Divider2.default, {
        key: i,
        inset: item.inset,
        style: item.style
      });
    } else {
      return _react2.default.createElement(_materialUi.ListItem, _extends({}, item, {
        key: i,
        value: item.value,
        onMouseDown: function onMouseDown(e) {
          if (e.button === 1) {
            var win = window.open('' + item.value, '_blank');
            win.focus();
          }
        },
        nestedItems: getNestedItems(item, i)
      }));
    }
  }

  return undefined;
}

var SelectableMenuList = function (_Component) {
  _inherits(SelectableMenuList, _Component);

  function SelectableMenuList() {
    _classCallCheck(this, SelectableMenuList);

    return _possibleConstructorReturn(this, _Component.apply(this, arguments));
  }

  SelectableMenuList.prototype.render = function render() {
    var _props = this.props,
        items = _props.items,
        onIndexChange = _props.onIndexChange,
        index = _props.index;


    var SelectableList = (0, _List.makeSelectable)(_materialUi.List);

    return _react2.default.createElement(
      SelectableList,
      {
        value: index,
        onChange: onIndexChange },
      items.filter(function (item) {
        return item.visible !== false;
      }).map(function (item, i) {
        return getItem(item, i);
      })
    );
  };

  return SelectableMenuList;
}(_react.Component);

;

SelectableMenuList.propTypes = process.env.NODE_ENV !== "production" ? {
  items: _propTypes2.default.array.isRequired,
  onIndexChange: _propTypes2.default.func.isRequired,
  index: _propTypes2.default.string.isRequired
} : {};

exports.default = SelectableMenuList;
module.exports = exports['default'];
