// muishift/lib/SelectField.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _MuiShift = require('./MuiShift');

var _MuiShift2 = _interopRequireDefault(_MuiShift);

var _TextField = require('@material-ui/core/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _InputAdornment = require('@material-ui/core/InputAdornment');

var _InputAdornment2 = _interopRequireDefault(_InputAdornment);

var _IconButton = require('@material-ui/core/IconButton');

var _IconButton2 = _interopRequireDefault(_IconButton);

var _ArrowDropDown = require('@material-ui/icons/ArrowDropDown');

var _ArrowDropDown2 = _interopRequireDefault(_ArrowDropDown);

var _ArrowDropUp = require('@material-ui/icons/ArrowDropUp');

var _ArrowDropUp2 = _interopRequireDefault(_ArrowDropUp);

var _Close = require('@material-ui/icons/Close');

var _Close2 = _interopRequireDefault(_Close);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var renderInput = function renderInput(props) {
  var downshiftProps = props.downshiftProps,
      rootProps = props.rootProps;
  var getInputProps = downshiftProps.getInputProps,
      isOpen = downshiftProps.isOpen,
      selectedItem = downshiftProps.selectedItem,
      openMenu = downshiftProps.openMenu,
      closeMenu = downshiftProps.closeMenu,
      clearSelection = downshiftProps.clearSelection;
  var inputProps = rootProps.inputProps,
      classes = rootProps.classes;

  var isDisabled = inputProps ? inputProps.disabled : undefined;
  var InputProps = getInputProps();

  return _react2.default.createElement(_TextField2.default, _extends({
    InputProps: _extends({
      onFocus: selectedItem ? undefined : openMenu,
      classes: {
        root: classes.inputRoot
      },
      endAdornment: _react2.default.createElement(
        _InputAdornment2.default,
        { position: 'end' },
        !!selectedItem && _react2.default.createElement(
          _IconButton2.default,
          {
            disabled: isDisabled,
            className: classes.closeButton,
            style: { width: 16 },
            onClick: clearSelection,
            tabIndex: -1 },
          _react2.default.createElement(_Close2.default, { style: { fontSize: 16 } })
        ),
        _react2.default.createElement(
          _IconButton2.default,
          { disabled: isDisabled, style: { width: 24 }, onClick: isOpen ? closeMenu : openMenu, tabIndex: -1 },
          isOpen ? _react2.default.createElement(_ArrowDropUp2.default, null) : _react2.default.createElement(_ArrowDropDown2.default, null)
        )
      )
    }, InputProps)
  }, inputProps));
};

var SelectField = function SelectField(props) {
  return _react2.default.createElement(_MuiShift2.default, _extends({ renderInput: renderInput }, props));
};

exports.default = SelectField;
module.exports = exports['default'];
