// muishift/lib/VirtualizedSelectField.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _SelectField = require('./SelectField');

var _SelectField2 = _interopRequireDefault(_SelectField);

var _reactVirtualized = require('react-virtualized');

var _Paper = require('@material-ui/core/Paper');

var _Paper2 = _interopRequireDefault(_Paper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var VirtualizedSelectField = function VirtualizedSelectField(props) {

  var renderMenu = function renderMenu(_ref) {
    var rootProps = _ref.rootProps,
        downshiftProps = _ref.downshiftProps,
        filteredItems = _ref.filteredItems;
    var height = props.height,
        rowHeight = props.rowHeight;
    var classes = rootProps.classes,
        renderSuggestion = rootProps.renderSuggestion;
    var highlightedIndex = downshiftProps.highlightedIndex;


    return _react2.default.createElement(
      _reactVirtualized.AutoSizer,
      null,
      function (_ref2) {
        var width = _ref2.width;
        return _react2.default.createElement(
          _Paper2.default,
          { className: classes.paper, style: { width: width }, square: true },
          _react2.default.createElement(_reactVirtualized.List, {
            width: width,
            scrollToIndex: highlightedIndex || 0,
            height: height,
            rowCount: filteredItems.length,
            rowHeight: rowHeight,
            rowRenderer: function rowRenderer(_ref3) {
              var key = _ref3.key,
                  index = _ref3.index,
                  style = _ref3.style;
              return _react2.default.createElement(
                'div',
                {
                  key: key,
                  style: _extends({}, style)
                },
                renderSuggestion({
                  rootProps: rootProps,
                  downshiftProps: downshiftProps,
                  suggestion: filteredItems[index],
                  index: index
                })
              );
            }
          })
        );
      }
    );
  };

  return _react2.default.createElement(_SelectField2.default, _extends({ renderMenu: renderMenu }, props));
};

VirtualizedSelectField.defaultProps = {
  height: 200,
  rowHeight: 48
};

exports.default = VirtualizedSelectField;
module.exports = exports['default'];
