// rmw-shell/lib/containers/AppLayout/AppLayout.js
'use strict';

exports.__esModule = true;
exports.AppLayout = undefined;

require('react-toastify/dist/ReactToastify.css');

var _Drawer = require('../../containers/Drawer');

var _Drawer2 = _interopRequireDefault(_Drawer);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Routes = require('../../containers/Routes');

var _Routes2 = _interopRequireDefault(_Routes);

var _reactToastify = require('react-toastify');

var _styles = require('@material-ui/core/styles');

var _messaging = require('../../utils/messaging');

var _reactIntl = require('react-intl');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = function styles(theme) {
  return {
    body: {
      position: 'absolute',
      top: 0,
      bottom: 0,
      left: 0,
      right: 0,
      width: '100%',
      height: '100%'
    },
    root: {
      flexGrow: 1,
      zIndex: 1,
      overflow: 'hidden',
      position: 'relative',
      display: 'flex',
      width: '100%'
    }
  };
};

var AppLayout = exports.AppLayout = function (_Component) {
  _inherits(AppLayout, _Component);

  function AppLayout() {
    _classCallCheck(this, AppLayout);

    return _possibleConstructorReturn(this, _Component.apply(this, arguments));
  }

  AppLayout.prototype.componentDidUpdate = function componentDidUpdate() {
    (0, _messaging.checkForUpdate)();
  };

  AppLayout.prototype.render = function render() {
    var classes = this.props.classes;

    return _react2.default.createElement(
      'div',
      { className: classes.body },
      _react2.default.createElement(
        'div',
        { className: classes.root },
        _react2.default.createElement(_Drawer2.default, null),
        _react2.default.createElement(_Routes2.default, null),
        _react2.default.createElement(_reactToastify.ToastContainer, null)
      )
    );
  };

  return AppLayout;
}(_react.Component);

exports.default = (0, _reactIntl.injectIntl)((0, _styles.withStyles)(styles, { withTheme: true })(AppLayout));
