// a2hs/lib/themeContext.js
'use strict';

exports.__esModule = true;
exports.A2HSContext = exports.initialState = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.withA2HS = withA2HS;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var initialState = exports.initialState = {
    deferredPrompt: function deferredPrompt() {},
    isAppInstallable: false,
    isAppInstalled: false
};

var A2HSContext = exports.A2HSContext = _react2.default.createContext(initialState);

// This function takes a component...
function withA2HS(Component) {
    // ...and returns another component...
    return function A2HSComponent(props) {
        // ... and renders the wrapped component with the context theme!
        // Notice that we pass through any additional props as well
        return _react2.default.createElement(
            A2HSContext.Consumer,
            null,
            function (context) {
                return _react2.default.createElement(Component, _extends({}, props, context));
            }
        );
    };
}

exports.default = A2HSContext;
