// rmw-shell/es/config/menuItems.js
import AccountBoxIcon from '@material-ui/icons/AccountBox';
import Brightness2 from '@material-ui/icons/Brightness2';
import Brightness7 from '@material-ui/icons/Brightness7';
import Icon from '@material-ui/core/Icon';
import LanguageIcon from '@material-ui/icons/Language';
import LockIcon from '@material-ui/icons/Lock';
import React from 'react';
import SettingsIcon from '@material-ui/icons/SettingsApplications';
import StyleIcon from '@material-ui/icons/Style';
import VerticalAlignBottomIcon from '@material-ui/icons/VerticalAlignBottom';
import allLocales from './locales';
import { themes as allThemes } from './themes';

var getMenuItems = function getMenuItems(props) {
  var locale = props.locale,
      updateTheme = props.updateTheme,
      switchNightMode = props.switchNightMode,
      updateLocale = props.updateLocale,
      intl = props.intl,
      themeSource = props.themeSource,
      auth = props.auth,
      isGranted = props.isGranted,
      deferredPrompt = props.deferredPrompt,
      isAppInstallable = props.isAppInstallable,
      isAppInstalled = props.isAppInstalled,
      isAuthMenu = props.isAuthMenu,
      handleSignOut = props.handleSignOut;


  var isAuthorised = auth.isAuthorised;

  var themeItems = allThemes.map(function (t) {
    return {
      value: undefined,
      visible: true,
      primaryText: intl.formatMessage({ id: t.id }),
      onClick: function onClick() {
        updateTheme(t.id);
      },
      leftIcon: React.createElement(StyleIcon, { style: { color: t.color } })
    };
  });

  var localeItems = allLocales.map(function (l) {
    return {
      value: undefined,
      visible: true,
      primaryText: intl.formatMessage({ id: l.locale }),
      onClick: function onClick() {
        updateLocale(l.locale);
      },
      leftIcon: React.createElement(LanguageIcon, null)
    };
  });

  if (isAuthMenu) {
    return [{
      value: '/my_account',
      primaryText: intl.formatMessage({ id: 'my_account' }),
      leftIcon: React.createElement(AccountBoxIcon, null)
    }, {
      value: '/signin',
      onClick: handleSignOut,
      primaryText: intl.formatMessage({ id: 'sign_out' }),
      leftIcon: React.createElement(LockIcon, null)
    }];
  }

  return [{
    value: '/dashboard',
    visible: isAuthorised,
    primaryText: intl.formatMessage({ id: 'dashboard' }),
    leftIcon: React.createElement(
      Icon,
      { className: 'material-icons' },
      'dashboard'
    )
  }, {
    visible: isAuthorised,
    primaryText: intl.formatMessage({ id: 'chats' }),
    primaryTogglesNestedList: true,
    leftIcon: React.createElement(
      Icon,
      { className: 'material-icons' },
      'chats'
    ),
    nestedItems: [{
      value: '/chats',
      visible: isAuthorised,
      primaryText: intl.formatMessage({ id: 'private' }),
      leftIcon: React.createElement(
        Icon,
        { className: 'material-icons' },
        'person'
      )
    }, {
      value: '/public_chats',
      visible: isAuthorised,
      primaryText: intl.formatMessage({ id: 'public' }),
      leftIcon: React.createElement(
        Icon,
        { className: 'material-icons' },
        'group'
      )
    }, {
      value: '/predefined_chat_messages',
      visible: isAuthorised,
      primaryText: intl.formatMessage({ id: 'predefined_messages' }),
      leftIcon: React.createElement(
        Icon,
        { className: 'material-icons' },
        'textsms'
      )
    }]
  }, {
    value: '/companies',
    visible: isGranted('read_companies'),
    primaryText: intl.formatMessage({ id: 'companies' }),
    leftIcon: React.createElement(
      Icon,
      { className: 'material-icons' },
      'business'
    )
  }, {
    value: '/tasks',
    visible: isAuthorised,
    primaryText: intl.formatMessage({ id: 'tasks' }),
    leftIcon: React.createElement(
      Icon,
      { className: 'material-icons' },
      'list'
    )
  }, {
    visible: isAuthorised,
    primaryTogglesNestedList: true,
    primaryText: intl.formatMessage({ id: 'firestore' }),
    leftIcon: React.createElement(
      Icon,
      { className: 'material-icons' },
      'flash_on'
    ),
    nestedItems: [{
      value: '/document',
      primaryText: intl.formatMessage({ id: 'document' }),
      leftIcon: React.createElement(
        Icon,
        { className: 'material-icons' },
        'flash_on'
      )
    }, {
      value: '/collection',
      primaryText: intl.formatMessage({ id: 'collection' }),
      leftIcon: React.createElement(
        Icon,
        { className: 'material-icons' },
        'flash_on'
      )
    }]
  }, {
    value: '/about',
    visible: isAuthorised,
    primaryText: intl.formatMessage({ id: 'about' }),
    leftIcon: React.createElement(
      Icon,
      { className: 'material-icons' },
      'info_outline'
    )
  }, {
    visible: isAuthorised, // In prod: isGranted('administration'),
    primaryTogglesNestedList: true,
    primaryText: intl.formatMessage({ id: 'administration' }),
    leftIcon: React.createElement(
      Icon,
      { className: 'material-icons' },
      'security'
    ),
    nestedItems: [{
      value: '/users',
      visible: isAuthorised, // In prod: isGranted('read_users'),
      primaryText: intl.formatMessage({ id: 'users' }),
      leftIcon: React.createElement(
        Icon,
        { className: 'material-icons' },
        'group'
      )
    }, {
      value: '/roles',
      visible: isGranted('read_roles'),
      primaryText: intl.formatMessage({ id: 'roles' }),
      leftIcon: React.createElement(
        Icon,
        { className: 'material-icons' },
        'account_box'
      )
    }]
  }, {
    divider: true,
    visible: isAuthorised
  }, {
    primaryText: intl.formatMessage({ id: 'settings' }),
    primaryTogglesNestedList: true,
    leftIcon: React.createElement(SettingsIcon, null),
    nestedItems: [{
      primaryText: intl.formatMessage({ id: 'theme' }),
      secondaryText: intl.formatMessage({ id: themeSource.source }),
      primaryTogglesNestedList: true,
      leftIcon: React.createElement(StyleIcon, null),
      nestedItems: themeItems
    }, {
      primaryText: intl.formatMessage({ id: 'language' }),
      secondaryText: intl.formatMessage({ id: locale }),
      primaryTogglesNestedList: true,
      leftIcon: React.createElement(LanguageIcon, null),
      nestedItems: localeItems
    }]
  }, {
    onClick: function onClick() {
      switchNightMode(!themeSource.isNightModeOn);
    },
    primaryText: intl.formatMessage({ id: themeSource.isNightModeOn ? 'day_mode' : 'night_mode' }),
    leftIcon: themeSource.isNightModeOn ? React.createElement(Brightness7, null) : React.createElement(Brightness2, null)
  }, {
    visible: isAppInstallable && !isAppInstalled,
    onClick: function onClick() {
      deferredPrompt.prompt();
    },
    primaryText: intl.formatMessage({ id: 'install' }),
    leftIcon: React.createElement(VerticalAlignBottomIcon, null)
  }];
};

export default getMenuItems;
