// firekit/lib/index.js
'use strict';

exports.__esModule = true;
exports.isInitialised = exports.getAllDocs = exports.getDoc = exports.getAllPaths = exports.getPath = exports.getAllInitializations = exports.getAllCols = exports.getCol = exports.hasError = exports.isLoading = exports.getAllLists = exports.getList = exports.unwatchAllPaths = exports.destroyPath = exports.unwatchPath = exports.watchPath = exports.unwatchAllLists = exports.destroyList = exports.unwatchList = exports.watchList = exports.unwatchAllDocs = exports.destroyDoc = exports.unwatchDoc = exports.watchDoc = exports.unwatchAllCols = exports.destroyCol = exports.unwatchCol = exports.watchCol = exports.authError = exports.authStateChanged = exports.watchAuth = exports.clearAllLoadings = exports.clearLoading = exports.logLoading = exports.clearAllErrors = exports.clearError = exports.logError = exports.clearMessage = exports.initMessaging = exports.unsubscribeConnection = exports.initConnection = exports.pathsReducer = exports.messagingReducer = exports.loadingsReducer = exports.listsReducer = exports.initializationReducer = exports.errorsReducer = exports.docsReducer = exports.connectionReducer = exports.collectionsReducer = exports.authReducer = exports.clearInitialization = undefined;

var _actions = require('./store/initialization/actions');

Object.defineProperty(exports, 'clearInitialization', {
  enumerable: true,
  get: function get() {
    return _actions.clearInitialization;
  }
});

var _reducer = require('./store/auth/reducer');

Object.defineProperty(exports, 'authReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_reducer).default;
  }
});

var _reducer2 = require('./store/collections/reducer');

Object.defineProperty(exports, 'collectionsReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_reducer2).default;
  }
});

var _reducer3 = require('./store/connection/reducer');

Object.defineProperty(exports, 'connectionReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_reducer3).default;
  }
});

var _reducer4 = require('./store/docs/reducer');

Object.defineProperty(exports, 'docsReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_reducer4).default;
  }
});

var _reducer5 = require('./store/errors/reducer');

Object.defineProperty(exports, 'errorsReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_reducer5).default;
  }
});

var _reducer6 = require('./store/initialization/reducer');

Object.defineProperty(exports, 'initializationReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_reducer6).default;
  }
});

var _reducer7 = require('./store/lists/reducer');

Object.defineProperty(exports, 'listsReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_reducer7).default;
  }
});

var _reducer8 = require('./store/loadings/reducer');

Object.defineProperty(exports, 'loadingsReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_reducer8).default;
  }
});

var _reducer9 = require('./store/messaging/reducer');

Object.defineProperty(exports, 'messagingReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_reducer9).default;
  }
});

var _reducer10 = require('./store/paths/reducer');

Object.defineProperty(exports, 'pathsReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_reducer10).default;
  }
});

var _actions2 = require('./store/connection/actions');

Object.defineProperty(exports, 'initConnection', {
  enumerable: true,
  get: function get() {
    return _actions2.initConnection;
  }
});
Object.defineProperty(exports, 'unsubscribeConnection', {
  enumerable: true,
  get: function get() {
    return _actions2.unsubscribeConnection;
  }
});

var _actions3 = require('./store/messaging/actions');

Object.defineProperty(exports, 'initMessaging', {
  enumerable: true,
  get: function get() {
    return _actions3.initMessaging;
  }
});
Object.defineProperty(exports, 'clearMessage', {
  enumerable: true,
  get: function get() {
    return _actions3.clearMessage;
  }
});

var _actions4 = require('./store/errors/actions');

Object.defineProperty(exports, 'logError', {
  enumerable: true,
  get: function get() {
    return _actions4.logError;
  }
});
Object.defineProperty(exports, 'clearError', {
  enumerable: true,
  get: function get() {
    return _actions4.clearError;
  }
});
Object.defineProperty(exports, 'clearAllErrors', {
  enumerable: true,
  get: function get() {
    return _actions4.clearAllErrors;
  }
});

var _actions5 = require('./store/loadings/actions');

Object.defineProperty(exports, 'logLoading', {
  enumerable: true,
  get: function get() {
    return _actions5.logLoading;
  }
});
Object.defineProperty(exports, 'clearLoading', {
  enumerable: true,
  get: function get() {
    return _actions5.clearLoading;
  }
});
Object.defineProperty(exports, 'clearAllLoadings', {
  enumerable: true,
  get: function get() {
    return _actions5.clearAllLoadings;
  }
});

var _actions6 = require('./store/auth/actions');

Object.defineProperty(exports, 'watchAuth', {
  enumerable: true,
  get: function get() {
    return _actions6.watchAuth;
  }
});
Object.defineProperty(exports, 'authStateChanged', {
  enumerable: true,
  get: function get() {
    return _actions6.authStateChanged;
  }
});
Object.defineProperty(exports, 'authError', {
  enumerable: true,
  get: function get() {
    return _actions6.authError;
  }
});

var _actions7 = require('./store/collections/actions');

Object.defineProperty(exports, 'watchCol', {
  enumerable: true,
  get: function get() {
    return _actions7.watchCol;
  }
});
Object.defineProperty(exports, 'unwatchCol', {
  enumerable: true,
  get: function get() {
    return _actions7.unwatchCol;
  }
});
Object.defineProperty(exports, 'destroyCol', {
  enumerable: true,
  get: function get() {
    return _actions7.destroyCol;
  }
});
Object.defineProperty(exports, 'unwatchAllCols', {
  enumerable: true,
  get: function get() {
    return _actions7.unwatchAllCols;
  }
});

var _actions8 = require('./store/docs/actions');

Object.defineProperty(exports, 'watchDoc', {
  enumerable: true,
  get: function get() {
    return _actions8.watchDoc;
  }
});
Object.defineProperty(exports, 'unwatchDoc', {
  enumerable: true,
  get: function get() {
    return _actions8.unwatchDoc;
  }
});
Object.defineProperty(exports, 'destroyDoc', {
  enumerable: true,
  get: function get() {
    return _actions8.destroyDoc;
  }
});
Object.defineProperty(exports, 'unwatchAllDocs', {
  enumerable: true,
  get: function get() {
    return _actions8.unwatchAllDocs;
  }
});

var _actions9 = require('./store/lists/actions');

Object.defineProperty(exports, 'watchList', {
  enumerable: true,
  get: function get() {
    return _actions9.watchList;
  }
});
Object.defineProperty(exports, 'unwatchList', {
  enumerable: true,
  get: function get() {
    return _actions9.unwatchList;
  }
});
Object.defineProperty(exports, 'destroyList', {
  enumerable: true,
  get: function get() {
    return _actions9.destroyList;
  }
});
Object.defineProperty(exports, 'unwatchAllLists', {
  enumerable: true,
  get: function get() {
    return _actions9.unwatchAllLists;
  }
});

var _actions10 = require('./store/paths/actions');

Object.defineProperty(exports, 'watchPath', {
  enumerable: true,
  get: function get() {
    return _actions10.watchPath;
  }
});
Object.defineProperty(exports, 'unwatchPath', {
  enumerable: true,
  get: function get() {
    return _actions10.unwatchPath;
  }
});
Object.defineProperty(exports, 'destroyPath', {
  enumerable: true,
  get: function get() {
    return _actions10.destroyPath;
  }
});
Object.defineProperty(exports, 'unwatchAllPaths', {
  enumerable: true,
  get: function get() {
    return _actions10.unwatchAllPaths;
  }
});

var _selectors = require('./store/lists/selectors');

Object.defineProperty(exports, 'getList', {
  enumerable: true,
  get: function get() {
    return _selectors.getList;
  }
});
Object.defineProperty(exports, 'getAllLists', {
  enumerable: true,
  get: function get() {
    return _selectors.getAllLists;
  }
});

var _selectors2 = require('./store/loadings/selectors');

Object.defineProperty(exports, 'isLoading', {
  enumerable: true,
  get: function get() {
    return _selectors2.isLoading;
  }
});

var _selectors3 = require('./store/errors/selectors');

Object.defineProperty(exports, 'hasError', {
  enumerable: true,
  get: function get() {
    return _selectors3.hasError;
  }
});

var _selectors4 = require('./store/collections/selectors');

Object.defineProperty(exports, 'getCol', {
  enumerable: true,
  get: function get() {
    return _selectors4.getCol;
  }
});
Object.defineProperty(exports, 'getAllCols', {
  enumerable: true,
  get: function get() {
    return _selectors4.getAllCols;
  }
});
Object.defineProperty(exports, 'getAllInitializations', {
  enumerable: true,
  get: function get() {
    return _selectors4.getAllInitializations;
  }
});

var _selectors5 = require('./store/paths/selectors');

Object.defineProperty(exports, 'getPath', {
  enumerable: true,
  get: function get() {
    return _selectors5.getPath;
  }
});
Object.defineProperty(exports, 'getAllPaths', {
  enumerable: true,
  get: function get() {
    return _selectors5.getAllPaths;
  }
});

var _selectors6 = require('./store/docs/selectors');

Object.defineProperty(exports, 'getDoc', {
  enumerable: true,
  get: function get() {
    return _selectors6.getDoc;
  }
});
Object.defineProperty(exports, 'getAllDocs', {
  enumerable: true,
  get: function get() {
    return _selectors6.getAllDocs;
  }
});

var _selectors7 = require('./store/initialization/selectors');

Object.defineProperty(exports, 'isInitialised', {
  enumerable: true,
  get: function get() {
    return _selectors7.isInitialised;
  }
});

var _reducer11 = _interopRequireDefault(_reducer);

var _reducer12 = _interopRequireDefault(_reducer2);

var _reducer13 = _interopRequireDefault(_reducer3);

var _reducer14 = _interopRequireDefault(_reducer4);

var _reducer15 = _interopRequireDefault(_reducer5);

var _reducer16 = _interopRequireDefault(_reducer6);

var _reducer17 = _interopRequireDefault(_reducer7);

var _reducer18 = _interopRequireDefault(_reducer8);

var _reducer19 = _interopRequireDefault(_reducer9);

var _reducer20 = _interopRequireDefault(_reducer10);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var firekitReducers = {
  auth: _reducer11.default,
  collections: _reducer12.default,
  connection: _reducer13.default,
  docs: _reducer14.default,
  errors: _reducer15.default,
  initialization: _reducer16.default,
  lists: _reducer17.default,
  loadings: _reducer18.default,
  messaging: _reducer19.default,
  paths: _reducer20.default
};

exports.default = firekitReducers;
