// rmw-shell/lib/components/ReduxFormFields/AvatarImageField.js
'use strict';

exports.__esModule = true;

var _IconButton = require('@material-ui/core/IconButton');

var _IconButton2 = _interopRequireDefault(_IconButton);

var _Icon = require('@material-ui/core/Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Avatar = require('../../components/ReduxFormFields/Avatar');

var _Avatar2 = _interopRequireDefault(_Avatar);

var _reduxForm = require('redux-form');

var _ImageCropDialog = require('../../containers/ImageCropDialog');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var AvatarImageField = function (_Component) {
  _inherits(AvatarImageField, _Component);

  function AvatarImageField(props) {
    _classCallCheck(this, AvatarImageField);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.handlePhotoUploadSuccess = function (snapshot) {
      var _this$props = _this.props,
          change = _this$props.change,
          name = _this$props.name;


      snapshot.ref.getDownloadURL().then(function (downloadURL) {
        change(name, downloadURL);
        _this.setState({ selected_avatar_image: undefined });
      });
    };

    _this.state = {
      selected_avatar_image: undefined
    };
    return _this;
  }

  AvatarImageField.prototype.render = function render() {
    var _this2 = this;

    var _props = this.props,
        altIconName = _props.altIconName,
        altIcon = _props.altIcon,
        disabled = _props.disabled,
        initialized = _props.initialized,
        intl = _props.intl,
        path = _props.path,
        uid = _props.uid,
        name = _props.name;


    return _react2.default.createElement(
      'div',
      { style: { margin: 20, display: 'flex', flexDirection: 'column', alignItems: 'center' } },
      _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(_reduxForm.Field, {
          name: name,
          style: { width: 120, height: 120, fontSize: 60 },
          component: _Avatar2.default,
          iconName: altIconName,
          icon: altIcon
        })
      ),
      _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          _IconButton2.default,
          {
            style: { width: '100%' },
            onClick: function onClick() {
              _this2.setState({ selected_avatar_image: 'true' });
            },
            disabled: disabled === true ? true : uid === undefined || !initialized,
            color: 'primary'
          },
          _react2.default.createElement(
            _Icon2.default,
            null,
            ' photo_camera '
          )
        )
      ),
      _react2.default.createElement(_ImageCropDialog.ImageCropDialog, {
        path: path + '/' + uid,
        fileName: name,
        onUploadSuccess: function onUploadSuccess(s) {
          _this2.handlePhotoUploadSuccess(s);
        },
        open: this.state.selected_avatar_image !== undefined,
        src: this.state.selected_avatar_image,
        handleClose: function handleClose() {
          _this2.setState({ selected_avatar_image: undefined });
        },
        title: intl.formatMessage({ id: 'change_photo' })
      })
    );
  };

  return AvatarImageField;
}(_react.Component);

AvatarImageField.propTypes = process.env.NODE_ENV !== "production" ? {
  uid: _propTypes2.default.string.isRequired,
  altIconName: _propTypes2.default.string,
  path: _propTypes2.default.string.isRequired,
  name: _propTypes2.default.string.isRequired
} : {};

exports.default = AvatarImageField;
module.exports = exports['default'];
